util/dvsim/Testplan.py

```diff
@@ -0,0 +1,686 @@
+                            t.replace("{" + item + "}", v) for t in resolved
+                        ]
+                else:
+                    resolved = [
+                        t.replace("{" + item + "}", value) for t in resolved
+                    ]
+
+            resolved_tests.extend(resolved)
+
+        self.tests = resolved_tests
+
+    def map_test_results(self, test_results):
+        """Map test results to tests against this testpoint.
+
+        Given a list of test results find the ones that match the tests listed
+        in this testpoint and buiild a structure. If no match is found, or if
+        self.tests is an empty list, indicate 0/1 passing so that it is
+        factored into the final total.
+        """
+        for tr in test_results:
+            assert isinstance(tr, TestResult)
+            if tr.name in self.tests:
+                tr.mapped = True
+                self.test_results.append(tr)
+
+        # Did we map all tests in this testpoint? If we are mapping the full
+        # testplan, then count the ones not found as "not run", i.e. 0 / 0.
+        tests_mapped = [tr.name for tr in self.test_results]
+        for test in self.tests:
+            if test not in tests_mapped:
+                self.test_results.append(TestResult(test))
+
+        # If no written tests were indicated for this testpoint, then reuse
+        # the testpoint name to count towards "not run".
+        if not self.tests:
+            self.test_results = [TestResult(self.name)]
+
+
+class Testplan():
+    """The full testplan
+
+    This comprises of Testpoints and Covergroups.
+    """
+
+    rsvd_keywords = ["import_testplans", "testpoints", "covergroups"]
+    element_cls = {'testpoint': Testpoint, 'covergroup': Covergroup}
+
+    @staticmethod
+    def _parse_hjson(filename):
+        """Parses an input file with HJson and returns a dict."""
+        try:
+            return hjson.load(open(filename, 'rU'))
+        except IOError as e:
+            print(f"IO Error when opening fie {filename}\n{e}")
+        except hjson.scanner.HjsonDecodeError as e:
+            print(f"Error: Unable to decode HJSON with file {filename}:\n{e}")
+        sys.exit(1)
+
+    @staticmethod
+    def _create_testplan_elements(kind, raw_dicts_list):
+        """Creates testplan elements from the list of raw dicts.
+
+        kind is either 'testpoint' or 'covergroup'.
+        raw_dicts_list is the list of raw dicts extracted from the HJson.
+        """
+        items = []
+        item_names = set()
+        for dict_entry in raw_dicts_list:
+            try:
+                item = Testplan.element_cls[kind](dict_entry)
+            except KeyError:
+                print(f"Error: {kind} arg is invalid")
+                sys.exit(1)
+            except ValueError as e:
+                print(e)
+                sys.exit(1)
+
+            if item.name in item_names:
+                print(f"Error: Duplicate {kind} item found with name: "
```

Comment:
If this is the place we will uniquify names, could you add a FIXME to that effect?

Reply:
I vaguely recall some mention of dealing with non-unique names getting uniquified using some path component? I was dreaming or that text needs revising.